### PR TITLE
feat(editor-client): search collection items

### PR DESCRIPTION
**Pull request by nicolaisirghi**

### CODE REVIEW

File: public/editor-client/src/types/global.d.ts
Comment (victor-pg) [inline public/editor-client/src/types/global.d.ts:1]: need to add also types for `collectionItems`

**Review by victor-pg — CHANGES_REQUESTED** [review]

File: public/editor-client/src/collectionItems/searchCollectionItems.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import { getConfig } from "src/config";
```
Comment (maxval1) [inline public/editor-client/src/collectionItems/searchCollectionItems.ts:1]: use relative import. `./go/to/path`

**Review by maxval1 — CHANGES_REQUESTED** [review]

File: public/editor-client/src/collectionItems/searchCollectionItems.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      const items = await getCollections(
+        { search, postType: [collectionId] },
+        config
+      );
+
+      const _items = items.map(({ title, permalink, ID }) => {
+        return {
+          title,
+          permalink,
+          id: read(ID) ?? ""
```
Comment (maxval1) [inline public/editor-client/src/collectionItems/searchCollectionItems.ts:35]: Filter all items and exclude items with invalid `ID`

File: public/editor-client/src/collectionItems/utils.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    search?: string;
+    include?: string[];
+    postType?: string[];
+    excludePostType?: string[];
+    abortSignal?: AbortSignal;
+  },
+  config: Config
+) => Promise<{ ID: number; title: string; permalink: string }[]>;
+
+export const getCollections: GetCollections = async (
```
Comment (maxval1) [inline public/editor-client/src/collectionItems/utils.ts:15]: move inside `api/index`

File: public/editor-client/src/collectionTypes/loadCollectionTypes.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { t } from "../utils/i18n";
+import { getCollectionTypes } from "./utils";
+
+export const loadCollectionTypes = {
+  async handler(res: Response<ChoicesSync>, rej: Response<string>) {
+    try {
+      const collectionTypes = getCollectionTypes();
+
+      if (!collectionTypes) {
+        throw new Error(t("Missing postLoopSources in config"));
```
Comment (maxval1) [inline public/editor-client/src/collectionTypes/loadCollectionTypes.ts:12]: call `rej()` don't need to create another error here

File: public/editor-client/src/collectionTypes/utils.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import { ChoicesSync } from "src/types/Choices";
+import { MValue } from "src/utils/types";
+
+export const getCollectionTypes = (): MValue<ChoicesSync> => {
+  const config = getConfig();
+
+  const collectionTypes = config?.collectionTypes;
+
+  if (collectionTypes && Array.isArray(collectionTypes)) {
+    return (collectionTypes ?? []).map(({ label, name }) => ({
```
Comment (maxval1) [inline public/editor-client/src/collectionTypes/utils.ts:11]: refactor, don't need `collectionTypes ?? [])`

File: public/editor-client/src/config.ts [hunk 6]
```diff
@@ -176,7 +200,15 @@ const reader = parseStrict<PLUGIN_ENV, Config>({
...
-  l10n: optional(pipe(Obj.readKey("l10n"), Obj.read, onNullish({})))
+  l10n: optional(pipe(Obj.readKey("l10n"), Obj.read, onNullish({}))),
+  collectionTypes: optional(
+    pipe(
+      Obj.readKey("collectionTypes"),
+      Arr.read,
+      collectionTypesReader,
+      throwOnNullish("")
+    )
+  )
```
Comment (maxval1) [inline public/editor-client/src/config.ts:211]: use `mPipe`

File: public/editor-client/src/config.ts [hunk 7]
```diff
@@ -67,6 +72,21 @@ const templatesReader = parseStrict<Record<string, unknown>, DefaultTemplates>({
...
+
+    return value.label && value.name;
+  });
+
+  if (isCollectionType) {
+    return arr as CollectionType[];
+  }
+  return undefined;
+};
+
```
Comment (maxval1) [inline public/editor-client/src/config.ts:89]:
```ts
const collectionTypesReader = (
  arr: Array<Record<string, unknown>>
): Array<CollectionType> => {
  return arr.filter((o): o is CollectionType => !!o.label && !!o.name);
};
```

**Review by maxval1 — CHANGES_REQUESTED** [review]

File: public/editor-client/src/collectionItems/searchCollectionItems.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+      const _items = items
+        .map(({ title, permalink, ID }) => {
+          return {
+            title,
+            permalink,
+            id: read(ID) ?? ""
+          };
+        })
+        .filter((item) => item.id);
```
Comment (maxval1) [inline public/editor-client/src/collectionItems/searchCollectionItems.ts:39]:
```ts
const _items = items
        .filter((item) => read(item.ID) );
        .map(({ title, permalink, ID }) => {
          return {
            title,
            permalink,
            id: ID
          };
        })
```

File: public/editor-client/src/config.ts [hunk 9]
```diff
@@ -176,7 +191,16 @@ const reader = parseStrict<PLUGIN_ENV, Config>({
     mPipe(Obj.readKey("api"), Obj.read, apiReader),
     throwOnNullish("Invalid: api")
   ),
-  l10n: optional(pipe(Obj.readKey("l10n"), Obj.read, onNullish({})))
+  l10n: optional(pipe(Obj.readKey("l10n"), Obj.read, onNullish({}))),
+  collectionTypes: pipe(
+    mPipe(
+      Obj.readKey("collectionTypes"),
+      Arr.read,
+      Arr.readUnknowToRecordReader,
```
Comment (maxval1) [inline public/editor-client/src/config.ts:199]: refactror

File: public/editor-client/src/utils/reader/array.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+  return undefined;
+};
+
+export const readUnknowToRecordReader = (
+  arr: unknown[]
+): Record<string, unknown>[] =>
+  arr.filter(
+    (o): o is Record<string, unknown> => typeof o === "object" && o != null
+  );
```
Comment (maxval1) [inline public/editor-client/src/utils/reader/array.ts:16]: ???

**Review by maxval1 — CHANGES_REQUESTED** [review]

**Pull request merged**